ZIP64 Support - Unzip(Reading)

Ckitakishi opened the pull request:
Fixes https://github.com/weichsel/ZIPFoundation/issues/118

# Changes proposed in this PR
* Data Descriptor
  * Convert types of `uncompressedSize` and `compressedSize` to a generic type which conforms to `BinaryInteger` (`Int64` or `UInt32`) 
* Apply zip64 data structures to entry extraction, entry removal and other unzip/read related code. 
* Add test cases:
  * Data Descriptor
  * Remove/Extract entry from zip64 archive
  * Zip/Unzip zip64 archive

# Tests performed
* All unit tests pass and coverage rate is 100%.
* SwiftLint errors will be resolved after the review is completed.
* Tested unarchive zip files over `4GB` on macOS, and it worked well.

# Further info for the reviewer
@weichsel Hi,  please review the second half of ZIP64 support when you have time~

Spec: https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT
Data Descriptor: if the zip64 extended information extra field is present for  the file the compressed and uncompressed sizes will be 8 byte values.  
```
crc-32                    4 bytes
compressed size           4 bytes/8 bytes
uncompressed size         4 bytes/8 bytes
```

# Open Issues
N/A

Ckitakishi committed:
Add `extraFields` property to avoid duplicate computations

Ckitakishi committed:
Add ZIP64 support for data descriptor

Ckitakishi committed:
Add convenience accessors to CD, they could return correct values no matter the zip is normal zip or zip64

Ckitakishi committed:
Fix all of the existing test cases as zip64 support was added to `checkIntegrity`

Ckitakishi committed:
Add test cases for zip/unzip zip64 format items.

Ckitakishi committed:
Add test cases for extract entry

Ckitakishi committed:
Reduce lines/paramters of methods

Ckitakishi committed:
Add test case to confirm if extra fields updated correctly when removing entry

Ckitakishi committed:
Fix a bug of zip64 extended information and improve test

Ckitakishi committed:
Add test cases to improve coverage and fix bugs
- ZIP64 data descriptor
- ZIP64 extra fields
- Extract entry with ZIP64 data descriptor

weichsel committed:
Move backing configuration related code into dedicated file

weichsel committed:
Move reading and writing helpers into dedicated file

weichsel committed:
Move entry data property access and initialization into extension

weichsel commented:
@Ckitakishi Sorry for the delayed reply. 

I started to look into this and this looks really great!
I fixed the remaining `swiftlint` issues.
Can you please give me [write permissions to your fork](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) so that I can directly push them.

Ckitakishi commented:
@weichsel No problem!

> I fixed the remaining swiftlint issues.

Thank you!! I have invited you to be a collaborator in this repository. I think you can directly push them now.

Ckitakishi committed:
Delete empty file

Ckitakishi commented:
[Update History] Deleted unused file at https://github.com/Ckitakishi/ZIPFoundation/pull/6/commits/91bb6aec7c1f0f566272d782b9a420a520ba6d3a

Ckitakishi commented:
Hi, @weichsel 
Do you think our review will still last for some time?

Since my team is planning to start testing ZIP64 support recently (many of our users are stuck with the ZIP64 limitation 😢 ), if you still need some time to review, we will test based on commits. 

Just a simple confirmation of your plan, feel free to take your time!

weichsel commented on Sources/ZIPFoundation/Archive+Helpers.swift:
```diff
@@ -0,0 +1,279 @@
+
+        let sizeOfCD = self.sizeOfCentralDirectory
+        let numberOfTotalEntries = self.totalNumberOfEntriesInCentralDirectory
+
```
```suggestion
```

weichsel commented on Sources/ZIPFoundation/Archive+Helpers.swift:
```diff
@@ -0,0 +1,279 @@
+                                    startOfEndOfCentralDirectory: Int64,
+                                    operation: ModifyOperation) throws -> EndOfCentralDirectoryStructure {
+        var record = self.endOfCentralDirectoryRecord
+
```
```suggestion
```

weichsel commented on Sources/ZIPFoundation/Archive+Helpers.swift:
```diff
@@ -0,0 +1,279 @@
+            ? UInt32.max
+            : UInt32(startOfCentralDirectory)
+        // ZIP64 End of Central Directory
+        var zip64eocd: ZIP64EndOfCentralDirectory?
```
nit: `zip64EOCD`

weichsel commented on Sources/ZIPFoundation/Archive+ZIP64.swift:
```diff
@@ -10,7 +10,7 @@
 
-/// The minimum version of zip64 format
+/// The minimum version of ZIP64 format
 let zip64Version = UInt16(45)
```
We could introduce a simple enum that currently only holds two cases:
```swift
public final class Archive {
// ... right after the `AccessMode` enum, we could add:

    enum Version: UInt16 {
        case v20 = 20 // minimum version for deflate compressed archives
        case v45 = 45 // minimum version for archives making use of ZIP64 extensions
    }

// ...
}
```

Instead of having a constant for version `45` here, and an inline magic number of `20` for non-ZIP64 archives, we could use the according `Version` case instead.

weichsel commented on Sources/ZIPFoundation/Archive.swift:
```diff
@@ -54,7 +54,8 @@ let memoryURLScheme = "memory"
 public final class Archive: Sequence {
     typealias LocalFileHeader = Entry.LocalFileHeader
-    typealias DataDescriptor = Entry.DataDescriptor
+    typealias DefaultDataDescriptor = Entry.DefaultDataDescriptor
```
I'd drop the `Default` prefix for the non-ZIP64 data descriptor here.
We also don't prefix e.g. `EndOfCentralDirectoryRecord` - which also has a ZIP64 variant.

weichsel commented on Tests/ZIPFoundationTests/ZIPFoundationFileManagerTests+ZIP64.swift:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    func testUnzipCompressedZIP64Item() {
+        // stored by zip 3.0 via command line: zip -0 -fz
+        //
+        // testUnzipCompressedZIP64Item.zip/
+        //   ├─ directory
+        //   ├─ testLink
+        //   ├─ nested
+        //     ├─ nestedLink
+        //     ├─ faust copy.txt
+        //     ├─ deep
+        //       ├─ another.random
+        //   ├─ faust.txt
+        //   ├─ empty
+        //   ├─ data.random
+        //   ├─ random.data
```
Great comment! 
Also a good idea to document the way how the test archive has been created.

weichsel commented on Tests/ZIPFoundationTests/ZIPFoundationProgressTests.swift:
```diff
@@ -212,7 +212,7 @@ extension ZIPFoundationTests {
         zipQueue.sync {
             XCTAssert(progress.fractionCompleted > 0.5)
-//            XCTAssert(archive.checkIntegrity())
+            XCTAssert(archive.checkIntegrity())
```
👍

weichsel commented on Tests/ZIPFoundationTests/ZIPFoundationReadingTests+ZIP64.swift:
```diff
@@ -0,0 +1,90 @@
+        }
+    }
+
+    // MARK: - helpers
```
nit: 

```suggestion
    // MARK: - Helpers
```

weichsel commented on Tests/ZIPFoundationTests/ZIPFoundationWritingTests+ZIP64.swift:
```diff
@@ -150,40 +150,40 @@ extension ZIPFoundationTests {
+        XCTAssert(entry.checksum == data.crc32(checksum: 0))
+        XCTAssert(archive.checkIntegrity())
+        XCTAssertEqual(entry.centralDirectoryStructure.relativeOffsetOfLocalHeader, UInt32.max)
+        XCTAssertEqual(entry.centralDirectoryStructure.extraFieldData.scanValue(start: 20), currentLFHOffset)
```
Thanks for re-adding those tests 👍

weichsel commented on Tests/ZIPFoundationTests/ZIPFoundationWritingTests+ZIP64.swift:
```diff
@@ -243,60 +246,80 @@ extension ZIPFoundationTests {
         let archive = self.archive(for: #function, mode: .update)
         guard let entry = archive["data1.random"] else {
-            XCTFail("Failed to add zip64 format entry to archive"); return
+            XCTFail("Failed to retrieve zip64 format entry from archive"); return
```
```suggestion
            XCTFail("Failed to retrieve ZIP64 format entry from archive"); return
```

weichsel commented on Sources/ZIPFoundation/Entry.swift:
```diff
@@ -59,17 +59,24 @@ public struct Entry: Equatable {
     }
 
-    struct DataDescriptor: DataSerializable {
+    struct DataDescriptor<T: BinaryInteger>: DataSerializable {
```
Nice use of generics. 
As far as I can see, we only use this pattern for `DataDescriptor` - is there a chance to use it for other ZIP/ZIP64 (e.g. EndOfCentralDirectoryRecord)? Or is this impossible due to different layout/members?

Ckitakishi replied:
I think it's hard to use Generic for other structure, the reason is as you said, other ZIP/ZIP64 structures have different layout/members. 😢 Other structures are not so GENERIC.

weichsel commented on Sources/ZIPFoundation/Entry.swift:
```diff
@@ -92,9 +99,16 @@ public struct Entry: Equatable {
         var usesUTF8PathEncoding: Bool { return (self.generalPurposeBitFlag & (1 << 11 )) != 0 }
         var isEncrypted: Bool { return (self.generalPurposeBitFlag & (1 << 0)) != 0 }
+        var isZIP64: Bool {
+            // If zip64 extended information is existing, try to treat cd as zip64 format
```
```suggestion
            // If ZIP64 extended information is existing, try to treat cd as ZIP64 format
```

weichsel commented on Tests/ZIPFoundationTests/ZIPFoundationTests.swift:
```diff
@@ -159,6 +159,20 @@ class ZIPFoundationTests: XCTestCase {
     }
+
+    // MARK: - ZIP64 Helpers
+    // It's not practical to create compressed files that exceed the size limit every time for test,
```
```suggestion

    // It's not practical to create compressed files that exceed the size limit every time for test,
```

weichsel commented on Tests/ZIPFoundationTests/ZIPFoundationWritingTests+ZIP64.swift:
```diff
@@ -26,10 +26,10 @@ extension ZIPFoundationTests {
         }
         guard let entry = archive[entryName] else {
-            XCTFail("Failed to add zip64 format entry to archive"); return
+            XCTFail("Failed to retrive zip64 format entry from archive"); return
```
```suggestion
            XCTFail("Failed to retrieve ZIP64 format entry from archive");
```

weichsel commented on Tests/ZIPFoundationTests/ZIPFoundationWritingTests+ZIP64.swift:
```diff
@@ -150,40 +150,40 @@ extension ZIPFoundationTests {
-//        XCTAssertEqual(entry.centralDirectoryStructure.relativeOffsetOfLocalHeader, UInt32.max)
-//        XCTAssertEqual(entry.centralDirectoryStructure.extraFieldData.scanValue(start: 4), currentLFHOffset)
+        guard let entry = archive[entryName] else {
+            XCTFail("Failed to retrive zip64 format entry from archive"); return
```
```suggestion
            XCTFail("Failed to retrieve ZIP64 format entry from archive"); return
```

weichsel commented on Tests/ZIPFoundationTests/ZIPFoundationWritingTests+ZIP64.swift:
```diff
@@ -150,40 +150,40 @@ extension ZIPFoundationTests {
-//        XCTAssertEqual(entry.centralDirectoryStructure.relativeOffsetOfLocalHeader, UInt32.max)
-//        XCTAssertEqual(entry.centralDirectoryStructure.extraFieldData.scanValue(start: 20), currentLFHOffset)
+        guard let entry = archive[entryName] else {
+            XCTFail("Failed to retrive zip64 format entry from archivee"); return
```
```suggestion
            XCTFail("Failed to retrieve ZIP64 format entry from archivee"); return
```

weichsel commented on Tests/ZIPFoundationTests/ZIPFoundationWritingTests+ZIP64.swift:
```diff
@@ -243,60 +246,80 @@ extension ZIPFoundationTests {
         let archive = self.archive(for: #function, mode: .update)
         guard let entry = archive["data2.random"] else {
-            XCTFail("Failed to add zip64 format entry to archive"); return
+            XCTFail("Failed to retrieve zip64 format entry from archive"); return
```
```suggestion
            XCTFail("Failed to retrieve ZIP64 format entry from archive"); return
```

weichsel commented on Tests/ZIPFoundationTests/ZIPFoundationWritingTests+ZIP64.swift:
```diff
@@ -243,60 +246,80 @@ extension ZIPFoundationTests {
+            XCTFail("Failed to remove entry from archive with error : \(error)")
+        }
+        XCTAssert(archive.checkIntegrity())
+        // Should delete zip64 extended information of data3.random as offset changed
```
nit: `zip64` should be replaced with `ZIP64` in those comments. (There are multiple occurrences of that string)

weichsel commented on Tests/ZIPFoundationTests/ZIPFoundationFileManagerTests+ZIP64.swift:
```diff
@@ -0,0 +1,139 @@
+    // MARK: - helpers
+
+    private func archiveZIP64Item(for testFunction: String, compressionMethod: CompressionMethod) throws {
+        mockIntMaxValues(int32Factor: 16, int16Factor: 16)
```
nit: please use explicit `self` when calling `mockIntMaxValues` (multiple occurrences)

weichsel reviewed (approved):
Thank you so much for this epic PR - Great work. 🎇
Found some minor code style issues. Will do a final round of testing after we merged that into the base PR.

Sorry for the super slow review - I was occupied with several other projects.

Ckitakishi committed:
Fix code style issues:
- Fix comment: change `zip64` to `ZIP64`
- Remove some empty lines
- Fix typo
- Use explicit `self` when calling mock methods

Ckitakishi committed:
Introduce a enum represent Archive Version

Ckitakishi committed:
Fix SwiftLint error automatically.

Ckitakishi commented:
> Sorry for the super slow review - I was occupied with several other projects.

That's fine!  I can totally relate to this.

> Found some minor code style issues. Will do a final round of testing after we merged that into the base PR.

Thanks for your thorough review, let me merge this PR into the base branch and continue discussion at https://github.com/weichsel/ZIPFoundation/pull/230
